### ISSUE TITLE
Prevent Now Playing Observable announcements until it's actually on-air

### ISSUE THREAD

**Issue by VxJasonxV** [ben-xo/sslscrobbler#14]
I'm using the Now Playing Observer to write out a file containing stream data, which is used by Nicecast in order to provide song metadata to our Icecast server. However, the metadata is written out as soon as the song is loaded. So when I'm shuffling through songs, a lot of erroneous data goes out.

I was watching the history pane in Serato DJ, and the data about the current song doesn't include a start time until the track actually goes on air, determined explicitly via both the crossfader, and the channel fader.

It seems to me that the underlying data provides a mechanism for knowing when a track is actually playing, but the reader naively treats a new song in the session file as being played.

The accuracy of the Now Playing observer would be dramatically improved if it evaluated this data (start time) before actually sending the announcements to the rest of the stack.

**Comment by ben-xo** [comment]
I feel criminal for not replying to this sooner. Hi @VxJasonxV , sorry it's been 8 years!

Aside from the very first track, the current version of SSLScrobbler shouldn't really show a song as now playing until all of the following conditions are met:

* the previous song is no longer "playing" OR there is no previous song; and
* the current song has been "playing" for more than `ScrobblerTrackModel::NOW_PLAYING_MIN` seconds.

NOW_PLAYING_MIN is currently set to 30 seconds, but you can adjust it. (This mainly avoids triggering loads of events during track-selection, but it also gives us a bit of mixing time / grace period for spoilers).

The previous song is considered to be no longer playing when it has been loaded for <previous song length> and therefore has "finished" playing. In practise that means it "finishes" from our model's perspective a bit before its natural end, due to not knowing how long it was cued up for before you really started to play it.

"There is no previous song" is triggered when you eject the previous song, so in general use, this is the condition which triggers a Now Playing event most often.

However, I'd love to use events based on audibility, if they were available. I don't use a controller, I still go oldschool with my Technics and DJM 800. 

I had a dream last night that it was time to upgrade that DJM 800…

**Comment by ben-xo** [comment]
(it would be easy to make NOW_PLAYING_MIN configurable at run time. It would probably break the few unit tests I wrote, though).

**Comment by kyds3k** [comment]
OK! The behavior I had noticed now makes sense, due to your previous song "no longer playing" criteria. I wish there was a more accurate way to tell - I often load the next track and have it sitting there for awhile before playing it, so I'm wondering if that throws it off.

Would it be possible to make the song "no longer playing" if it's been ejected from  the deck? That sounds like something I could remember to do and give more accuracy . . .

**Comment by ben-xo** [comment]
> Would it be possible to make the song "no longer playing" if it's been ejected from the deck? That sounds like something I could remember to do and give more accuracy . . .

It already does that; that's how I use it :)

**Comment by kyds3k** [comment]
> > Would it be possible to make the song "no longer playing" if it's been ejected from the deck? That sounds like something I could remember to do and give more accuracy . . .
> 
> It already does that; that's how I use it :)

I just loaded these 2 tracks:

![image](https://user-images.githubusercontent.com/1071985/116544524-efe3bc00-a8f7-11eb-8549-4ed725595930.png)

the track on the right was at 2:20 or so, with over 2 minutes left . . . I loaded the left hand track in and started cueing it up to mix and the Now Playing changed to the left hand track after about 20 seconds, even though the currently playing track hadn't ended yet.

Terminal output: 

`1: NEW        [--:--] [Love and Rockets - Mirror People '88 ((Single Edit))] [prev: Cocteau Twins - Beatrix]
2: PLAYING    [--:--] [Xmal Deutschland - Mondlicht] [prev: The Cure - The Hanging Garden]
Date: 2021-04-29 11:30:38 Memory Usage: 2,854,336 bytes
2021-04-29 11:30:38 WARNING: SSLTrack - Guessed MP3 length 236 seconds from file.
2021-04-29 11:30:38 INFO: NowPlayingModel - enqueued track Love and Rockets - Mirror People '88 ((Single Edit))
2021-04-29 11:30:38 INFO: DmcaAlerter - Checking if track would break any Mixcloud DMCA limits
2021-04-29 11:30:38 INFO: DmcaAlerter - Nah, we're fine!
2021-04-29 11:31:07 INFO: ScrobblerTrackModel - Love and Rockets - Mirror People '88 ((Single Edit)) passed now playing point
2021-04-29 11:32:35 INFO: ScrobblerTrackModel - Love and Rockets - Mirror People '88 ((Single Edit)) passed scrobble point
2021-04-29 11:32:46 INFO: ScrobblerTrackModel - Xmal Deutschland - Mondlicht passed end point
2021-04-29 11:34:34 INFO: ScrobblerTrackModel - Love and Rockets - Mirror People '88 ((Single Edit)) passed end point
2021-04-29 11:35:25 INFO: DiffMonitor - Ignoring update to older session file /Users/adamdarby/Music/_Serato_/History/Sessions/5961.session`

I started the whole thing with `php ./historyreader.php  -J 8080`

Any ideas?

**Comment by ben-xo** [comment]
Sure so `2021-04-29 11:32:46 INFO: ScrobblerTrackModel - Xmal Deutschland - Mondlicht passed end point` means you must have had that track loaded onto that deck for more than the length of the entire length of the song, so it assumes you must have listened to the whole thing by now. Do you load your tracks onto the deck minutes in advance? I tend to only load a track I'm going to play about a minute before I play it, so I haven't had problems with SSLScrobbler thinking the song has ended when it's only half way through. (It usually thinks its ended about 80% of the way through if i don't eject it first).

Without any code changes, the solution would be to keep the track off the deck a bit longer until you are ready to play it. But I could also change the code to make it wait longer e.g. song length x1.5 before treating it as ended, or make that configurable to your taste.

**Comment by kyds3k** [comment]
IIRC, I had the Xmal song loaded in there for only 20 seconds, and then it switched over. I kind of do a fly by the seat of my pants style of DJing, so I'll load in like 3-4 tracks before settling on one, and then it sits there for awhile. A configurable wait period would be awesome!

**Comment by ben-xo** [comment]
> IIRC, I had the Xmal song loaded in there for only 20 seconds, and then it switched over. I kind of do a fly by the seat of my pants style of DJing, so I'll load in like 3-4 tracks before settling on one, and then it sits there for awhile. A configurable wait period would be awesome!

Yeah that's how I do it too. But it's actually probably that the song before was loaded for so long that it thought it must have finished by now, which is quite common if you load your first track a lot earlier than subsequent tracks. It might just be that, although I've got some things in the works to try to help with that situation too. I'll keep you updated
